pkg/logging/config.go

```diff
@@ -50,13 +50,11 @@ func NewLogger(configJSON string, levelOverride string) (*zap.SugaredLogger, zap
 		}
 	}
 
-	logger, err2 := loggingCfg.Build()
+	logger, err2 := loggingCfg.Build(zap.Fields(zap.String("fallback-logger", "true")))
```

Comment:
This looks good, thanks! I wonder how this could be tested. Is it possible to extract the list of fields for a particular logger? Could the logger's writer be swapped out (maybe with https://godoc.org/go.uber.org/zap#WrapCore) and then a loged message checked for the correct string?

Reply:
Let me look into this more. I could not find an obvious way to test it.

Reply:
@grantr 

Tried couple of things and this is what I found so far.

1. Use file or stdout buffer
```
	logger, _ := NewLogger("some invalid JSON here", "")
       // Trying to set logger 
        temp, _ := ioutil.TempFile("", "zapcore-test")
		nop := zap.WrapCore(func(z zapcore.Core) zapcore.Core {
		return zapcore.NewCore(
			zapcore.NewJSONEncoder(testEncoderConfig()),
			temp,
			zapcore.DebugLevel,
		)
	})
        
        logger.With(nop).Info("test log!!!")
        logged, err := ioutil.ReadFile(temp.Name())
	fmt.Printf("Logger info : %s", logged.String())
```
Tried to use stdout/ file to dump the logs and read it back after logging. Sugared logger panic with Unknown key. I did try to Desugar the logger but could not extract the key.

2. I tried to redirect Stdout(using os.Pipe) into buffer and set the core to forward logs to STDOUT. No logs came through stdout.

3. I tried to use RegisterHook func but I did not have luck verifying anything from core object about the key.  

I am not sure if I am approaching the problem in a wrong way. If you have any suggestions , then please let me know. 
FYI This looks very promising: https://godoc.org/go.uber.org/zap/zaptest/observer 
It captures logs in memory. This library is used extensively in sugar logger unit testing. I am not sure whether its worth vendoring dependency to test this feature.

Reply:
It does seem worth testing this particular feature since it would fail silently. That library seems totally reasonable to vendor for this case.

Here's what I expected would be possible, but I haven't tested to make sure it's actually possible. :smile:

```go
logger, _ := NewLogger("some invalid JSON here", "")
var buf bytes.Buffer
bufcore := zap.WrapCore(func(zapcore.Core) zapcore.Core {
  return zapcore.NewCore(
    zapcore.NewJsonEncoder(testEncoderConfig()),
    zapcore.AddSync(buf),
    zapcore.DebugLevel,
  )
}

logger.With(bufcore).Info("test log")
m, err := regexp.MatchString("fallback-logger",buf.String())
if err != nil {
  t.Error(err)
  }
  if !b {
    t.Errorf("fallback-logger string not in output")
  }
}
```

Reply:
In the above example `sugar` logger does not add add anything to buffer. I tried to `Desugar` the logger but then I get only key `msg` and value in buffer. Rest of metadata is lost.

Reply:
Dang :( What if you call `logger.Sync()` before inspecting the buffer?

Reply:
I did call `Sync()` but did not make any difference to output. 
I tried with `zaptest` package as well

```
core, logged := observer.New(zapcore.DebugLevel)
logger.With(core).Debug("test log")
logger.Sync()
if logged.Len() < 1 {
	t.Errorf("expected at least one log line")
}
```

I am not having any luck with testing this feature :(

Reply:
@shashwathi I noticed that since the fallback logger is a Production logger, it samples log lines. In my testing that caused my test logs to be ignored. I recommend changing the fallback logger to be either a development logger or a production logger with sampling disabled.

I think what we're trying to do is not possible with zap's interfaces. Specifically, **it is impossible to wrap a core while keeping its structured fields.** Fields set on the logger using the `With` option are never exposed once set, and `WrapCore` doesn't help because it can't retrieve the encoder of the existing core. `RegisterHooks` doesn't receive the logger's fields, so it's also a dead end.

I would really like to test this feature, but I'm fine with not having it tested for now since the `fallback-logger` string is clearly visible in the log lines emitted by the test. I think the best remaining option is [Testable Examples](https://blog.golang.org/examples) but I haven't had much luck with that yet.

/lgtm

Reply:
@grantr : I tried `Testable Examples` as well and it was another dead-end.

Reply:
@shashwathi Thanks for your efforts and sorry about all the dead ends. Maybe we can fall back to an e2e test (but I don't want to hold up this PR for that).

@mdemirhan what do you think about changing the fallback logger to disable sampling? Fallback is already a confusing situation for the user, and I think sampling would make it even more confusing.